File: Server~/src/tools/refreshAssetsTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'refresh_assets';
const toolDescription = 'Refreshes the Unity AssetDatabase to import any changed assets (equivalent to Assets -> Refresh).';

const paramsSchema = z.object({});

/**
 * Creates and registers the Refresh Assets tool with the MCP server
 *
 * @param server The MCP server instance to register with
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param logger The logger instance for diagnostic information
 */
export function registerRefreshAssetsTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  // Register this tool with the MCP server
  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

/**
 * Handles refresh assets tool requests
 *
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param params The parameters for the tool
 * @returns A promise that resolves to the tool execution result
 * @throws McpUnityError if the request to Unity fails
 */
async function toolHandler(mcpUnity: McpUnity, params: z.infer<typeof paramsSchema>): Promise<CallToolResult> {
  // Send to Unity
  const response = await mcpUnity.sendRequest({
    method: toolName,
    params
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to refresh assets'
    );
  }

  return {
    content: [
      {
        type: 'text',
        text: response.message || 'Successfully refreshed assets.'
      }
    ]
  };
}

File: Editor/Tools/RefreshAssetsTool.cs
using System;
using McpUnity.Unity;
using McpUnity.Utils;
using UnityEditor;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for refreshing the AssetDatabase to import any changed assets.
    /// </summary>
    public class RefreshAssetsTool : McpToolBase
    {
        public RefreshAssetsTool()
        {
            Name = "refresh_assets";
            Description = "Refreshes the AssetDatabase to import any changed assets, equivalent to clicking 'Assets -> Refresh' in the Editor.";
        }
        
        /// <summary>
        /// Execute the RefreshAssets tool synchronously on the main thread
        /// </summary>
        /// <param name="parameters">Tool parameters as a JObject</param>
        /// <returns>Result status of the asset refresh as a JObject</returns>
        public override JObject Execute(JObject parameters)
        {
            McpLogger.LogInfo("[MCP Unity] Refreshing assets...");
            
            try
            {
                // AssetDatabase.Refresh syncs all changes on the disk to the editor
                AssetDatabase.Refresh();
                
                return new JObject
                {
                    ["success"] = true,
                    ["type"] = "text",
                    ["message"] = "Successfully refreshed Unity AssetDatabase."
                };
            }
            catch (Exception ex)
            {
                McpLogger.LogError($"[MCP Unity] Failed to refresh assets: {ex.Message}");
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Failed to refresh assets: {ex.Message}", 
                    "tool_execution_error"
                );
            }
        }
    }
}
